planner/core/rule_join_elimination.go

```diff
@@ -173,38 +163,57 @@ func (o *outerJoinEliminator) isDuplicateAgnosticAgg(p LogicalPlan) (_ bool, col
 			return false, nil
 		}
 		for _, expr := range aggDesc.Args {
-			if col, ok := expr.(*expression.Column); ok {
-				cols = append(cols, col)
-			} else {
-				return false, nil
-			}
+			cols = append(cols, expression.ExtractColumns(expr)...)
 		}
 	}
 	return true, cols
 }
 
-func (o *outerJoinEliminator) doOptimize(p LogicalPlan, aggCols []*expression.Column, parentSchema *expression.Schema) (LogicalPlan, error) {
+func (o *outerJoinEliminator) doOptimize(p LogicalPlan, aggCols []*expression.Column, parentCols []*expression.Column) (LogicalPlan, error) {
+	var err error
+	var isEliminated bool
+	for join, isJoin := p.(*LogicalJoin); isJoin; join, isJoin = p.(*LogicalJoin) {
+		p, isEliminated, err = o.tryToEliminateOuterJoin(join, aggCols, parentCols)
+		if err != nil {
+			return p, err
+		}
+		if !isEliminated {
+			break
+		}
+	}
+
+	var colsInSchema []*expression.Column
+	switch x := p.(type) {
+	case *LogicalProjection:
+		for _, expr := range x.Exprs {
+			colsInSchema = append(colsInSchema, expression.ExtractColumns(expr)...)
+		}
+	case *LogicalAggregation:
+		colsInSchema = x.groupByCols
+		for _, aggDesc := range x.AggFuncs {
+			for _, expr := range aggDesc.Args {
+				colsInSchema = append(colsInSchema, expression.ExtractColumns(expr)...)
```

Comment:
It would change the contents of `x.groupByCols`. Would it bring some potential bugs?

Why not continue using `parentCols`? BTW, can we follow the same rule to maintain `parentCols` in the column pruning rule?

Reply:
1. The `colsInSchema` actually means the `parentCols`. When meeting `LogicalAgg`, it should be the `groupByCols` addictive with columns in `expr`.
2. The way maintaining `parentCols` in column pruning doesn't fit the situation. Because the `parentCols` is checked by the outer child of the plan, rather than itself in the column pruning rule.